src/WebJobs.Script.WebHost/DependencyInjection/JobHostScopedServiceProviderFactory.cs

```diff
@@ -29,12 +41,316 @@ public IServiceCollection CreateBuilder(IServiceCollection services)
 
         public IServiceProvider CreateServiceProvider(IServiceCollection containerBuilder)
         {
+            GetExpectedServices().Validate(containerBuilder);
+
             if (_provider == null)
             {
                 _provider = new JobHostServiceProvider(containerBuilder, _rootProvider, _rootScopeFactory);
             }
 
             return _provider;
         }
+
+        private static ExpectedServices GetExpectedServices()
+        {
+            var expected = new ExpectedServices();
+
+            expected.ExpectNone<IScriptEventManager>();
+            expected.ExpectNone<IEventGenerator>();
+
+            expected.Expect<ILoggerFactory, ScriptLoggerFactory>();
+
+            expected.Expect<IMetricsLogger, WebHostMetricsLogger>();
+
+            expected.Expect<IEventCollectorFactory>("Microsoft.Azure.WebJobs.Logging.EventCollectorFactory");
+
+            expected.ExpectCollection<IEventCollectorProvider>()
+                .Expect<FunctionInstanceLogCollectorProvider>()
+                .Expect("Microsoft.Azure.WebJobs.Logging.FunctionResultAggregatorProvider");
+
+            expected.ExpectCollection<IHostedService>()
+                .Expect<JobHostService>()
+                .Expect<JobHostService>("Microsoft.Azure.WebJobs.Hosting.OptionsLoggingService")
+                .Expect<PrimaryHostCoordinator>()
+                .Expect<FunctionsSyncService>()
+                .Expect<HttpInitializationService>()
+                .Expect<FileMonitoringService>()
+                .Expect<LanguageWorkerConsoleLogService>();
+
+            expected.ExpectSubcollection<ILoggerProvider>()
+                .Expect<UserLogMetricsLoggerProvider>()
+                .ExpectFactory<Startup>()
+                .Expect<SystemLoggerProvider>();
+
+            return expected;
```

Comment:
@fabiocav -- this is the API I'm creating. It handles a bunch of the different scenarios (internal types, no matches, subcollections, collections, etc). Will improve doc comments as well in next iteration, but I think you understand most scenarios.

Reply:
👍